File: frame_karaoke/utils/audio_utils.py
"""Audio processing utilities."""

import struct
import logging
from typing import List

logger = logging.getLogger(__name__)


def combine_audio_chunks(chunks: List[bytes]) -> bytes:
    """
    Combine multiple audio chunks into a single buffer.

    Args:
        chunks: List of audio data chunks

    Returns:
        Combined audio data
    """
    return b''.join(chunks)


def trim_audio_buffer(
    audio_data: bytes,
    max_duration_seconds: float,
    sample_rate: int = 16000
) -> bytes:
    """
    Trim audio buffer to a maximum duration.

    Keeps the most recent audio (end of buffer).

    Args:
        audio_data: Audio data to trim
        max_duration_seconds: Maximum duration in seconds
        sample_rate: Audio sample rate (default: 16000 Hz)

    Returns:
        Trimmed audio data
    """
    bytes_per_second = sample_rate * 2  # 16-bit mono = 2 bytes per sample
    max_bytes = int(max_duration_seconds * bytes_per_second)

    if len(audio_data) <= max_bytes:
        return audio_data

    # Keep only the last max_bytes
    return audio_data[-max_bytes:]


def calculate_audio_duration(
    audio_length: int,
    sample_rate: int = 16000
) -> float:
    """
    Calculate audio duration from buffer length.

    Args:
        audio_length: Length of audio data in bytes
        sample_rate: Audio sample rate (default: 16000 Hz)

    Returns:
        Duration in seconds
    """
    bytes_per_second = sample_rate * 2  # 16-bit mono
    return audio_length / bytes_per_second


def create_wav_header(data_length: int, sample_rate: int = 16000) -> bytes:
    """
    Create WAV file header for PCM audio.

    Format: 16-bit PCM, mono channel

    Args:
        data_length: Length of audio data in bytes
        sample_rate: Audio sample rate (default: 16000 Hz)

    Returns:
        44-byte WAV header
    """
    # Constants for 16-bit mono PCM
    num_channels = 1
    bits_per_sample = 16
    byte_rate = sample_rate * num_channels * (bits_per_sample // 8)
    block_align = num_channels * (bits_per_sample // 8)

    # Build header (44 bytes total)
    header = bytearray()

    # RIFF chunk descriptor (12 bytes)
    header += b'RIFF'                                    # ChunkID
    header += struct.pack('<I', data_length + 36)       # ChunkSize (file size - 8)
    header += b'WAVE'                                    # Format

    # fmt sub-chunk (24 bytes)
    header += b'fmt '                                    # Subchunk1ID
    header += struct.pack('<I', 16)                     # Subchunk1Size (16 for PCM)
    header += struct.pack('<H', 1)                      # AudioFormat (1 = PCM)
    header += struct.pack('<H', num_channels)           # NumChannels
    header += struct.pack('<I', sample_rate)            # SampleRate
    header += struct.pack('<I', byte_rate)              # ByteRate
    header += struct.pack('<H', block_align)            # BlockAlign
    header += struct.pack('<H', bits_per_sample)        # BitsPerSample

    # data sub-chunk (8 bytes)
    header += b'data'                                    # Subchunk2ID
    header += struct.pack('<I', data_length)            # Subchunk2Size

    return bytes(header)


def create_wav_buffer(audio_data: bytes, sample_rate: int = 16000) -> bytes:
    """
    Create complete WAV file buffer from PCM audio data.

    Args:
        audio_data: Raw PCM audio data
        sample_rate: Audio sample rate (default: 16000 Hz)

    Returns:
        Complete WAV file as bytes (header + data)
    """
    header = create_wav_header(len(audio_data), sample_rate)
    return header + audio_data


def validate_audio_data(audio_data: bytes) -> bool:
    """
    Validate audio data.

    Args:
        audio_data: Audio data to validate

    Returns:
        True if valid, False otherwise
    """
    if not audio_data or len(audio_data) == 0:
        logger.warning("Audio data is empty")
        return False

    # Check if length is reasonable (at least 0.1 seconds at 16kHz)
    min_bytes = int(0.1 * 16000 * 2)  # 0.1 seconds
    if len(audio_data) < min_bytes:
        logger.warning(f"Audio data too short: {len(audio_data)} bytes")
        return False

    return True

File: flutter_karaoke/lib/utils/audio_utils.dart
import 'dart:typed_data';

/// Audio processing utilities

/// Combine multiple audio chunks into a single buffer
Uint8List combineAudioChunks(List<Uint8List> chunks) {
  if (chunks.isEmpty) return Uint8List(0);

  final totalLength = chunks.fold<int>(0, (sum, chunk) => sum + chunk.length);
  final combined = Uint8List(totalLength);

  var offset = 0;
  for (final chunk in chunks) {
    combined.setRange(offset, offset + chunk.length, chunk);
    offset += chunk.length;
  }

  return combined;
}

/// Trim audio buffer to a maximum duration
Uint8List trimAudioBuffer(
  Uint8List audioData,
  double maxDurationSeconds,
  int sampleRate,
) {
  final bytesPerSecond = sampleRate * 2; // 16-bit mono = 2 bytes per sample
  final maxBytes = (maxDurationSeconds * bytesPerSecond).floor();

  if (audioData.length <= maxBytes) {
    return audioData;
  }

  // Keep only the last maxBytes
  return Uint8List.sublistView(audioData, audioData.length - maxBytes);
}

/// Calculate audio duration from buffer length
double calculateAudioDuration(int audioLength, int sampleRate) {
  final bytesPerSecond = sampleRate * 2; // 16-bit mono
  return audioLength / bytesPerSecond;
}

/// Create WAV file header for PCM audio
Uint8List createWavHeader(int dataLength, int sampleRate) {
  final header = ByteData(44);

  // Constants for 16-bit mono PCM
  const numChannels = 1;
  const bitsPerSample = 16;
  final byteRate = sampleRate * numChannels * (bitsPerSample ~/ 8);
  final blockAlign = numChannels * (bitsPerSample ~/ 8);

  // RIFF chunk descriptor (12 bytes)
  header.setUint8(0, 'R'.codeUnitAt(0));
  header.setUint8(1, 'I'.codeUnitAt(0));
  header.setUint8(2, 'F'.codeUnitAt(0));
  header.setUint8(3, 'F'.codeUnitAt(0));
  header.setUint32(4, dataLength + 36, Endian.little);
  header.setUint8(8, 'W'.codeUnitAt(0));
  header.setUint8(9, 'A'.codeUnitAt(0));
  header.setUint8(10, 'V'.codeUnitAt(0));
  header.setUint8(11, 'E'.codeUnitAt(0));

  // fmt sub-chunk (24 bytes)
  header.setUint8(12, 'f'.codeUnitAt(0));
  header.setUint8(13, 'm'.codeUnitAt(0));
  header.setUint8(14, 't'.codeUnitAt(0));
  header.setUint8(15, ' '.codeUnitAt(0));
  header.setUint32(16, 16, Endian.little); // Subchunk1Size (16 for PCM)
  header.setUint16(20, 1, Endian.little); // AudioFormat (1 = PCM)
  header.setUint16(22, numChannels, Endian.little); // NumChannels
  header.setUint32(24, sampleRate, Endian.little); // SampleRate
  header.setUint32(28, byteRate, Endian.little); // ByteRate
  header.setUint16(32, blockAlign, Endian.little); // BlockAlign
  header.setUint16(34, bitsPerSample, Endian.little); // BitsPerSample

  // data sub-chunk (8 bytes)
  header.setUint8(36, 'd'.codeUnitAt(0));
  header.setUint8(37, 'a'.codeUnitAt(0));
  header.setUint8(38, 't'.codeUnitAt(0));
  header.setUint8(39, 'a'.codeUnitAt(0));
  header.setUint32(40, dataLength, Endian.little);

  return header.buffer.asUint8List();
}

/// Create complete WAV file buffer from PCM audio data
Uint8List createWavBuffer(Uint8List audioData, int sampleRate) {
  final header = createWavHeader(audioData.length, sampleRate);
  final wavBuffer = Uint8List(header.length + audioData.length);

  wavBuffer.setRange(0, header.length, header);
  wavBuffer.setRange(header.length, wavBuffer.length, audioData);

  return wavBuffer;
}

/// Validate audio data
bool validateAudioData(Uint8List audioData) {
  if (audioData.isEmpty) {
    return false;
  }

  // Check if length is reasonable (at least 0.1 seconds at 16kHz)
  const minBytes = 3200; // 0.1 seconds at 16kHz * 2 bytes per sample
  if (audioData.length < minBytes) {
    return false;
  }

  return true;
}
